diff --git a/src/osgUtil/SceneView.cpp b/src/osgUtil/SceneView.cpp
index d6763eafcf8..526ea790d21 100644
--- a/src/osgUtil/SceneView.cpp
+++ b/src/osgUtil/SceneView.cpp
@@ -313,10 +313,6 @@ void SceneView::init()
 {
     _initCalled = true;
 
-    // force the initialization of the OpenGL extension string
-    // to try and work around a Windows NVidia driver bug circa Oct 2006.
-    osg::isGLExtensionSupported(_renderInfo.getState()->getContextID(),"");
-
     if (_camera.valid() && _initVisitor.valid())
     {
         _initVisitor->reset();
